File: c++/leetcode/111. Minimum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  int minDepth(TreeNode *root)
  {
    if (!root)
      return 0;

    int _min = 0;

    if (root->left && root->right)
    {
      int left = 1 + minDepth(root->left);
      int right = 1 + minDepth(root->right);

      _min = min(left, right);
    }
    else if (root->right)
      _min = 1 + minDepth(root->right);
    else if (root->left)
      _min = 1 + minDepth(root->left);
    else
      _min = 1;

    return _min;
  }
};

File: javascript/leetcode/111. Minimum Depth of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (!root) return 0;
  let _min = 0;

  const left = 1 + minDepth(root.left);
  const right = 1 + minDepth(root.right);

  if (root.left && root.right) _min = Math.min(left, right);
  else if (root.left) _min = left;
  else if (root.right) _min = right;
  else _min = 1;

  return _min;
};
